### ISSUE TITLE
swtpm_cert fails test test_swtpm_cert: Certificate file has wrong size.

### ISSUE THREAD

**Issue by gsx** [stefanberger/swtpm#2]
Hi,

My build fails test `test_swtpm_cert`. Relevant excerpt from `tests/test-suite.log`:

```
FAIL: test_swtpm_cert
=====================

Error: Certificate file has wrong size.
       Expected: 1224;  found: 1216
FAIL test_swtpm_cert (exit status: 1)
```

I checked and it is not just the first certificate that's problematic. All four certificates generated have different sizes than expected: always 4 less than the test case expects, except in the first case where it's 8 less than it should be. The generated certificates seem to be ok by simple visual inspection (attached the first one for reference; please excuse the file extension: [cert1.txt](https://github.com/stefanberger/swtpm/files/60680/cert1.txt)), and the output of `openssl x509 -text -in cert1.txt` (see below) seems mostly plausible to me as well. However, there are some strange characters (terminal escape sequences?) in the SAN field, the like of which I have not seen before. Is that intentional?

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
    Signature Algorithm: sha1WithRSAEncryption
        Issuer: CN=swtpm-localca
        Validity
            Not Before: Dec 14 00:34:10 2015 GMT
            Not After : Dec 11 00:34:10 2025 GMT
        Subject: 
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:b9:dd:a8:30:72:9d:e5:8f:9f:5b:ed:2b:3b:93:
                    94:ad:4e:c5:af:b9:c3:90:b8:9a:33:37:25:0c:bc:
                    57:5c:fc:8f:31:f7:ff:d3:f0:5f:41:55:07:6f:7d:
                    16:05:38:1c:d2:81:b7:f1:47:b8:01:15:4e:4f:89:
                    ee:52:9f:e3:6e:ae:50:f7:95:61:85:0e:5b:63:03:
                    7e:da:ac:bb:39:0e:a3:fc:d0:37:e6:74:fb:17:9e:
                    3c:5a:fe:31:21:4d:78:a7:56:ca:44:cc:6c:f2:54:
                    21:b5:14:20:ed:e5:48:31:0c:92:b0:8a:51:3c:cc:
                    62:fd:0e:f4:5d:cf:65:46:f6:e8:65:be:6a:66:1d:
                    04:5d:1c:47:b6:0b:42:8d:11:dc:97:cb:9f:35:ee:
                    7c:38:5b:b2:03:20:93:4b:01:5f:80:14:e8:fb:19:
                    85:1c:2a:f3:07:e1:e6:46:48:c1:42:17:5e:40:b6:
                    06:15:dc:49:4f:db:09:ea:5d:5a:6f:32:73:b6:5a:
                    24:1e:3c:f3:0c:c4:49:b9:fb:3f:90:0d:1e:d4:be:
                    96:7b:32:b1:6f:95:a1:d7:32:db:fa:14:3e:aa:1c:
                    20:17:55:61:17:f7:0f:ae:e5:d7:7f:83:67:05:d0:
                    54:05:36:1a:d5:87:1a:32:16:1f:a5:a1:23:4c:fa:
                    b4:97
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Extended Key Usage: 
                rsaesOaep
            X509v3 Subject Alternative Name: 
                URI:0*0^L^F^Eg<81>^E^B^A^L^CIBM0^L^F^Eg<81>^E^B^B^L^CIBM0^L^F^Eg<81>^E^B^C^L^CIBM
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Authority Key Identifier: 
                keyid:67:2D:C2:5E:CB:61:DF:51:D3:72:30:A2:58:D0:06:A0:8D:8F:69:91

            X509v3 Key Usage: critical
                Key Encipherment
    Signature Algorithm: sha1WithRSAEncryption
         65:65:8a:d9:c4:b7:d8:d0:56:4e:1d:ca:63:73:01:89:33:7b:
         8b:80:52:38:df:61:58:a5:0e:10:ba:f2:e7:48:fc:cc:7d:11:
         42:3f:41:2c:5e:1e:6f:c3:5c:62:0e:57:1f:6f:da:52:f3:a1:
         84:8e:57:c9:51:9c:92:a3:99:53:c8:b3:02:8c:99:21:1d:04:
         89:dd:ab:33:70:0a:b4:49:26:cf:c3:56:0b:15:cd:1b:67:72:
         22:55:16:2d:90:ea:f4:62:06:8f:15:59:eb:11:88:9d:61:c1:
         17:76:83:ad:b3:69:eb:c4:5b:5f:0e:1d:35:28:10:36:8b:1f:
         dd:47:c0:79:3d:59:67:36:0f:4d:4d:38:83:e6:54:78:a9:2b:
         38:0a:3a:c9:1b:f3:8c:35:56:28:09:15:f9:c5:68:40:e6:9d:
         f1:c3:cc:cd:7c:43:61:19:c0:3d:68:58:2c:75:30:c6:e7:2b:
         44:1e:c1:69:b3:5a:a4:83:03:03:88:ab:bc:b3:f2:2d:dc:f0:
         11:7e:e6:76:85:74:06:0f:ad:ce:15:94:78:bd:a5:e4:c3:c2:
         cf:10:c1:a1:57:39:44:f1:1c:8f:74:a7:8d:22:6e:d5:bf:29:
         30:84:f0:e2:c8:c3:f4:4d:bf:ea:3a:11:c2:fd:d9:72:7b:8a:
         9e:60:2b:a5:7d:89:cb:ef:74:5b:fd:31:06:a2:4d:a0:4d:67:
         5a:63:51:83:19:5f:3d:ef:d3:59:da:fb:6e:68:21:e2:9a:3e:
         f1:6b:1c:1f:da:c4:03:e6:dc:8b:68:5c:b8:57:ba:9c
-----BEGIN CERTIFICATE-----
MIIDVjCCAg6gAwIBAgIBATANBgkqhkiG9w0BAQUFADAYMRYwFAYDVQQDEw1zd3Rw
bS1sb2NhbGNhMB4XDTE1MTIxNDAwMzQxMFoXDTI1MTIxMTAwMzQxMFowADCCASIw
DQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALndqDByneWPn1vtKzuTlK1Oxa+5
w5C4mjM3JQy8V1z8jzH3/9PwX0FVB299FgU4HNKBt/FHuAEVTk+J7lKf426uUPeV
YYUOW2MDftqsuzkOo/zQN+Z0+xeePFr+MSFNeKdWykTMbPJUIbUUIO3lSDEMkrCK
UTzMYv0O9F3PZUb26GW+amYdBF0cR7YLQo0R3JfLnzXufDhbsgMgk0sBX4AU6PsZ
hRwq8wfh5kZIwUIXXkC2BhXcSU/bCepdWm8yc7ZaJB488wzESbn7P5ANHtS+lnsy
sW+Vodcy2/oUPqocIBdVYRf3D67l13+DZwXQVAU2GtWHGjIWH6WhI0z6tJcCAwEA
AaOBkjCBjzAUBgNVHSUEDTALBgkqhkiG9w0BAQcwNwYDVR0RBDAwLoYsMCowDAYF
Z4EFAgEMA0lCTTAMBgVngQUCAgwDSUJNMAwGBWeBBQIDDANJQk0wDAYDVR0TAQH/
BAIwADAfBgNVHSMEGDAWgBRnLcJey2HfUdNyMKJY0AagjY9pkTAPBgNVHQ8BAf8E
BQMDByAAMA0GCSqGSIb3DQEBBQUAA4IBMQBlZYrZxLfY0FZOHcpjcwGJM3uLgFI4
32FYpQ4QuvLnSPzMfRFCP0EsXh5vw1xiDlcfb9pS86GEjlfJUZySo5lTyLMCjJkh
HQSJ3aszcAq0SSbPw1YLFc0bZ3IiVRYtkOr0YgaPFVnrEYidYcEXdoOts2nrxFtf
Dh01KBA2ix/dR8B5PVlnNg9NTTiD5lR4qSs4CjrJG/OMNVYoCRX5xWhA5p3xw8zN
fENhGcA9aFgsdTDG5ytEHsFps1qkgwMDiKu8s/It3PARfuZ2hXQGD63OFZR4vaXk
w8LPEMGhVzlE8RyPdKeNIm7VvykwhPDiyMP0Tb/qOhHC/dlye4qeYCulfYnL73Rb
/TEGok2gTWdaY1GDGV8979NZ2vtuaCHimj7xaxwf2sQD5tyLaFy4V7qc
-----END CERTIFICATE-----
```

Code was compiled from commit 86b6ffca753 with the openssl crypto backend (openssl version 1.0.2.e from Arch Linux). Please let me know if I can provide any more information which may be useful.

**Comment by stefanberger** [comment]
Thank you for the error report and running the tests. 

I think the tests for the size of the created certificate are too strict and different versions of the GnuTLS library may add additional fields or data to the certificate, resulting in different sizes. So I will remove the tests for the size.

The strange characters in the SAN field are due to ASN.1 representation of information passed via the command line. The implementation follows the TCG document "TCG Credential Profiles  
For TPM Family 1.2; Level 2" creating a Unified Certificate. In section 3.5.9 the contents of Subject Alternative Names is described and is supposed to contain the description of Platform and TPM Manufacturer: "Depending on the type of certificate (see EKU  discussion in 3.5.16 below) this field MUST contain different lists of names in the SubjectAltName  extension (which is a sequence of GeneralNames identified by an OID.)" 

http://www.trustedcomputinggroup.org/files/static_page_files/A55529C5-1A4B-B294-D0A5A400E1EDE13A/Credential_Profiles_V1.2_Level2_Revision8.pdf

**Comment by stefanberger** [comment]
I pushed an update now that should solve this problem. Please pull and try again and let me know if there are other problems.

**Comment by gsx** [comment]
Thank you for the swift response, Stefan. The problem is now solved, the test suite passes. (Funnily enough, I was about to report the "return value ignored" problem as well, since it was causing problems when compiling with _FORTIFY_SOURCE, but you fixed it before I got to it.)

Also, thanks for the explanation and reference regarding the SAN field. I have not yet familiarized myself too closely with the TCG specifications, but I guess it's high time to.
